runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -837,11 +837,15 @@ public void onProcessingTime(InternalTimer<ByteBuffer, TimerData> timer) {
 
   // allow overriding this in ExecutableStageDoFnOperator to set the key context
   protected void fireTimerInternal(Object key, TimerData timerData) {
-    fireTimer(timerData);
+    if (key instanceof java.nio.ByteBuffer) {
+      key = FlinkKeyUtils.decodeKey((ByteBuffer) key, keyCoder);
```

Comment:
@mxm Can you take a quick look and verify that this key handling is correct here?

Reply:
The key is always stored as ByteBuffer. This is also reflected in the method calls to this method. Please change the signature of the method to be 
```
void fireTimerInternal(ByteBuffer key, TimerData timerdata)
```

Reply:
@mxm, upon debugging, I noticed that timer.getKey() returns String, Integer, as well as ByteBuffer. That's why when I changed the method signature, DoFnOperatorTest failed with ClassCastException.

Reply:
Isn't `timer.getKey()` a new interface? Keys are always internally stored encoded as ByteBuffer. The keys itself can of course be of any type. The test might not instrument the code correctly.

Reply:
@mxm should I use a different interface to get the encoded key? Isn't timer.getKey was supposed to return ByteBuffer always?

Reply:
The key is always encoded at this point. The issue is that some unit tests use a non-encoded key which is incorrect. We have to fix this to change the method signature to `ByteBuffer` here. As a workaround you could use `keyedStateInternals.getCurrentKey()` which will return a decoded version of the key.

Reply:
If you do, please add a `// TODO` comment alongside with it.

Reply:
```keyStateInternal.getKey()``` internally assumes that the key is of type ```ByteBuffer``` and tries to decode it. Since we have some test that uses decoded key, we get ClassCastException here as well. To prevent it I have added a ```key instanceof java.nio.ByteBuffer``` check. This will let me know when to decode, as a workaround. @mxm

Reply:
Yes, it assumes that the key is of type ByteBuffer because the key type we pass to Flink is always `ByteBuffer`. We shouldn't be adding `instanceof` checks to the main code in order for tests to pass. Let's adapt the tests instead. You need to ensure that the `KeySelector` used in the tests returns `ByteBuffer`. I can help you in case you get stuck.

Reply:
Thanks for providing the lead. I am on it.